File: house-robber/dp.go
package main

import "fmt"

func rob(nums []int) int {
	prev_max, running_max := 0, 0

	// [prev_max, running_max, v, v+1]
	for _, v := range nums {
		tmp := running_max
		if v+prev_max > tmp {
			tmp = v + prev_max
		}

		prev_max = running_max
		running_max = tmp
	}

	return running_max
}

func main() {
	fmt.Println(rob([]int{2, 7, 9, 3, 1}))
}

File: house-robber-ii/dp.py
from typing import List
def rob(self, nums: List[int]) -> int:
    def helper (nums: List[int]) -> int:
        prev2, prev = 0, 0
        for val in nums:
            tmp = max(prev2 + val, prev)
            prev2 = prev
            prev = tmp
        return prev
    return max(helper(nums[1:]), helper(nums[:-1]), nums[0])

print(rob(None, [10]))